src/vm/virtualMachine.ts

```diff
@@ -125,381 +57,292 @@ export class VirtualMachine {
             }
             break;
-        }
-      default:
-        return undefined;
-    }
-  }
-
-  deduceOp1(
-    instruction: Instruction,
-    dst: SegmentValue | undefined,
-    op0: SegmentValue | undefined
-  ): SegmentValue | undefined {
-    if (instruction.opcode !== Opcode.AssertEq) {
-      return undefined;
-    }
-    switch (instruction.resLogic) {
-      // If the result logic is op1, then dst = op1.
-      case ResLogic.Op1:
-        return dst;
-
-      case ResLogic.Add:
-        // dst := res = op0 + op1
-        // op1 = op0 - dst
-        if (dst !== undefined && op0 !== undefined) {
-          return dst.sub(op0);
-        }
-        break;
-      case ResLogic.Mul:
-        // dst := res = op0 * op1
-        // op1 = dst / op0
-        if (dst !== undefined && op0 !== undefined) {
-          try {
-            if (!isFelt(dst)) {
-              throw new Error();
-            }
-            return dst.div(op0);
-          } catch (err) {
-            return undefined;
-          }
+          case ResLogic.Unused:
+            res = undefined;
+            break;
+          default:
+            throw new Error();
         }
         break;
     }
-    return undefined;
-  }
 
-  // Compute the result of an instruction based on result logic
-  // for the instruction.
-  computeRes(
-    instruction: Instruction,
-    op0: SegmentValue,
-    op1: SegmentValue
-  ): SegmentValue | undefined {
-    switch (instruction.resLogic) {
-      case ResLogic.Op1:
-        return op1;
-      case ResLogic.Add:
-        return op0.add(op1);
-      case ResLogic.Mul:
-        if (!isFelt(op0)) {
-          throw new ExpectedFelt();
-        }
-        return op0.mul(op1);
-      case ResLogic.Unused:
-        return undefined;
-    }
+    if (op0 !== undefined) this.memory.assertEq(op0Addr, op0);
+    if (op1 !== undefined) this.memory.assertEq(op1Addr, op1);
+    if (dst !== undefined) this.memory.assertEq(dstAddr, dst);
+
+    return {
+      op0,
+      op1,
+      res,
+      dst,
+    };
   }
 
-  // Deduce the destination of an instruction. We can only deduce
-  // for assert eq and call instructions.
-  deduceDst(
+  /** Perform checks for Call opcode cases.
+   *
+   * Compute op0 if not retrieved from memory
+   */
+  private validateOperandsCallOpcode(
     instruction: Instruction,
-    res: SegmentValue | undefined
-  ): SegmentValue | undefined {
-    switch (instruction.opcode) {
-      // For an assert equal, we have dst := res
-      case Opcode.AssertEq:
-        return res;
-      // For a call instruction, we have dst := fp
-      case Opcode.Call:
-        return this.fp;
-      default:
-        return undefined;
-    }
+    op0: SegmentValue | undefined,
+    op1: SegmentValue | undefined
+  ): { op0: SegmentValue; op1: SegmentValue } {
+    const nextPc = this.pc.add(instruction.size());
+    if (op0 === undefined) {
+      op0 = nextPc;
+    } else if (!op0.eq(nextPc)) throw new InvalidOp0();
```

Comment:
this could be made more explicit that the validation is about 

1. op0 === nextPc
2. op1 !== undefined

so I'd move away the op0 resolution

```
op0 = op0 === undefined ? nextPc : op0;
```

possibly compute nextPc, nextAp and nextFp before so that it's done and here
and write only the `if then throw` that could possibly be inlined as well